### DISCUSSION TITLE
Keep transparency when converting Tiff to PNG in all cases

### DISCUSSION THREAD

**Discussion by nyroDev** [ImageMagick/ImageMagick#4946] [Help]
I managing an online tool where images are uplaoded from various users.
In order to work with uplaoded files, I have to convert them into PNGs.

When doing so, I want to be as close as possible as what they expect.
Sometimes I encounter TIFF files which looks transparent on Photoshop, but I cannot find a way to converti it to PNG with the transparency.

I dig it a while, tested a lot of combination of options, but I cannot find a way to do that.

I finally found that the main difference was the "save transparency" option on Photoshop.
When it's checked, everthing works fine.
When it's not, I end with a white image.

I created 2 exemples :
[test_transparency.zip](https://github.com/ImageMagick/ImageMagick/files/8244093/test_transparency.zip)
Zip contains : 
- test_transparent.tif: Saved with "Save transparency" checked
- test_noTransparent.tif: Saved without "Save transparency" checked

When simply doing somehting like:

```
convert test_noTransparent.tif -background none -flatten +repage -strip test_noTransparent.png
convert test_transparent.tif -background none -flatten +repage -strip test_transparent.png
```

I ended up with :  
![test_transparent](https://user-images.githubusercontent.com/265662/158158593-dbb13768-bd3a-405f-af80-cee236e35b21.png)
![test_noTransparent](https://user-images.githubusercontent.com/265662/158158616-80b8ea17-881c-42fa-8350-e172f25571fe.png)

As you can see, the second one is not transparent.
In Photoshop, it's exactly the same source file, only the "save transparency" changes.

What I'm looking for is a way to correctly convert both of these files.
I could add a previous step in order to identify if I'm facing a case like that.

Note that I simplify the test case as much as possible, but I could have Tiff files with more layers.
I could also face tiff files without background transparecy too...

Bonus: I'm also interested to understand why identify these tiff files show 2 layers.

Thanks in advance.

**Comment by fmw42** [comment]
When you save a file in Photoshop, you must tell it to save the transparency, if you want to export to non-Photoshop systems. I do not know if there is any way to save it currently without "saving" the transparency.

**Comment by snibgo** [comment]
@nyroDev: You seem to think that the tif files contain one image each. In fact, each contains two images:
```
magick test_transparent.tif info:

test_transparent.tif[0] TIFF 200x200 200x200+0+0 8-bit TrueColor sRGB 23448B 0.000u 0:00.002
test_transparent.tif[1] TIFF 105x99 105x99+30+26 8-bit sRGB 23448B 0.063u 0:00.026
```
If you want the first image, I suggest you explicitly use a suffix [0]. Or [1] for the second image.

**Reply by nyroDev** [reply]
I created a pretty simple example for this help request.
I opened Photoshop, and added a single layer with the colored square.
In Photoshop or The Gimp, there is only 1 layer shown.

the size of my image is 200x200.
But If I'm using only the first layer [0], I'm getting a full white square.

My understanding after a lot of reading:
The first layer :
- Set the image complete dimensions (200x200)
- should be interpreted as transparent, but it's done in a way Imagemagick doesn't know (because it's not standard) -> The result is a white 200x200

If I want to do that, I need a way to detect when the first layer is in this specific case.
For now, I didn't manage to find a proper way to that.

I'll then merge all layers except the first one, and use the dimension of the first layer.
(I have real use cases with more than 2 real layer)

**Comment by snibgo** [comment]
When you open test_notransparent.tif in Photoshop, does it show transparency? I don't have PS so can't test this.

Gimp and exiftool and IM can't see any transparency in test_notransparent.tif.

**Reply by nyroDev** [reply]
yes it is
![Capture d'écran de 2022-03-15 09-55-45](https://user-images.githubusercontent.com/265662/158341607-99eff705-413e-4d46-9223-fb70725a491f.png)
.

**Comment by fmw42** [comment]
@dlemstra  I opened the TIFF in PS and can see that the transparency is not checked when I try to save again as TIFF. Then I closed the TIFF and reopened it. The (background) transparency is still in the image. So background transparency is saved in a way that PS knows it is there even though the checkbox is not checked and PS can re-display the transparency. Perhaps it is in the xmp profile?

**Comment by snibgo** [comment]
We can write XMP metadata to a text file:
```
magick test_noTransparent.tif[0] x.xmp
```
There is no transparency data there. It may be in one of the other profiles:
```
  Profiles:
    Profile-8bim: 4802 bytes
    Profile-psd:additional-info: 260 bytes
    Profile-xmp: 14015 bytes
```
IM contains code in coders\psd.c to process "psd:additional-info", but this seems limited to simply copying the data.

We can use "-define psd:additional-info=all" and "-define psd:preserve-opacity-mask=true" (see https://imagemagick.org/script/formats.php) when reading test_noTransparent.tif, and write a new PSD file, but that doesn't seem useful.

Perhaps some software exists somewhere that can read and process (and even explain) "8bim" and "psd:additional-info" profiles.

**Reply by nyroDev** [reply]
After trying to extract 8bim profiles and open them as text file, I can see that there is difference.  
For transparent.tiff, there is a mention of "Transparence" while on the same place, there is only garbish character or something like.

I tried to find a way to extract information from 8bim in order to be human readable, but i cannot find anything.
Is there a tool somewhere or a lib that can do that?
